src/Workspaces/Remote/ServiceHub/Shared/RemoteEndPoint.cs

```diff
@@ -122,6 +122,22 @@ public async Task<T> InvokeAsync<T>(string targetName, IReadOnlyList<object> arg
             }
         }
 
+        public Task<TResult> InvokeAsync<TResult>(
+            string targetName, IReadOnlyList<object> arguments, Func<Stream, CancellationToken, TResult> funcWithDirectStream, CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfFalse(_startedListening);
+            cancellationToken.ThrowIfCancellationRequested();
+
+            try
+            {
+                return Extensions.InvokeAsync(_rpc, targetName, arguments, funcWithDirectStream, cancellationToken);
+            }
+            catch (Exception ex) // no when since Extensions.InvokeAsync already recorded it
+            {
+                throw CreateSoftCrashException(ex, cancellationToken);
```

Comment:
reading this code tends to confuse me.  because it means cancellation *looks* like it would throw soft-crash.  either it does... and that's odd.  or it doesn't, in which case it's not clear here why that would be the case.

also, all the different try/catch/throw patterns we have in OOP are completely impossible for me to rationalize.  they all seem to be different and follow different unclear rules.

Reply:
Are you looking at the latest version of the PR. This code doesn't exist anymore.

Reply:
i am not.  you mentioned revieweing a commit at a time. so that's what i was doing :)

Reply:
Well, yes, that was if you want to see a sequence of smaller refactorings that led to the resulting PR. But that approach obviously means that I made more changes to the same code later.